### DISCUSSION TITLE
Consider toowoxx/packer for uploading the snapshot.

### DISCUSSION THREAD

**Discussion by Roxedus** [hcloud-talos/terraform-hcloud-talos#68] [General]
With version `v0.6.0-alpha.2` of the Talos provider, it exposes a schematic resource. paring this with building the snapshot directly with Terraform makes bootstrapping the snapshot easier.

The provider does state the artifact does not store well in the Terraform state, but I think this is not too big of a deal. (One might be able to do keep it in state with data providers)

You can today abuse the fact that Terraform grabs a local copy of the repository, and use the packer module today.
```hcl
data "talos_image_factory_extensions_versions" "this" {
  talos_version = var.talos_version
  filters = {
    names = [
      "iscsi-tools",
      "tailscale",
    ]
  }
}

resource "talos_image_factory_schematic" "this" {
  schematic = yamlencode(
    {
      customization = {
        systemExtensions = {
          officialExtensions = data.talos_image_factory_extensions_versions.this.extensions_info.*.name
        }
      }
    }
  )
}

resource "packer_image" "talos" {
  file = "${path.module}/.terraform/modules/terraform-hcloud-talos/_packer/talos-hcloud.pkr.hcl"

  triggers = {
    talos_version = var.talos_version
    schematic     = talos_image_factory_schematic.this.id
    sha           = filesha256("${path.module}/.terraform/modules/terraform-hcloud-talos/_packer/talos-hcloud.pkr.hcl")
  }

  variables = {

    talos_version   = var.talos_version
    server_location = var.datacenter_name
    image_url_arm   = "https://factory.talos.dev/image/${talos_image_factory_schematic.this.id}/${var.talos_version}/hcloud-arm64.raw.xz"
    image_url_x86   = "https://factory.talos.dev/image/${talos_image_factory_schematic.this.id}/${var.talos_version}/hcloud-amd64.raw.xz"
  }
}
```

**Comment by mrclrchtr** [comment]
That sounds very interesting! I was also planning to add autoscaling at some point. You probably need the stored “finished” snapshot for this. But I'll have to take a closer look at the concept first.

Having the option to deploy without snapshot is definitely great. I'll have a look at that.

Unfortunately, the problem at the moment is a lack of time :/

**Comment by mrclrchtr** [comment]
I took a look at it and it could actually be a cool solution, but we would have to use [this provider](https://github.com/toowoxx/terraform-provider-packer), correct? The repository has very few stars and I would really like to only use popular providers here, because of security reasons.

**Comment by mrclrchtr** [comment]
I created https://github.com/hcloud-talos/terraform-provider-imager
